src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,234 @@
+        final Class<?> returnType = target.type().returnType();
+        Objects.requireNonNull(stateName);
+
+        if (!(returnType.equals(int.class) || returnType.equals(long.class))) {
+            throw illegalArgDoesNot(target, "return an int or a long");
+        }
+        if (target.type().parameterType(0) != MemorySegment.class) {
+            throw illegalArgDoesNot(target, "have a MemorySegment as the first parameter");
+        }
+
+        // (int | long)(int | long)
+        final MethodHandle returnFilter = RETURN_FILTERS
+                .get(returnType)
+                .get(stateName);
+        if (returnFilter == null) {
+            throw new IllegalArgumentException("Unknown state name: " + stateName);
+        }
+
+        // (MemorySegment, C*)(int | long) -> (C*)(int | long)
+        target = MethodHandles.collectArguments(target, 0, ACQUIRE_MH);
+
+        // (C*)(int | long) -> (C*)(int | long)
+        return MethodHandles.filterReturnValue(target, returnFilter);
+    }
+
+    // Used reflectively via ACQUIRE_MH
+    private static MemorySegment acquireCaptureStateSegment() {
+        MemorySegment segment = TL.get();
+        if (segment == null) {
+            TL.set(segment = malloc());
+        }
+        return segment;
+    }
+
+    // Used reflectively via INT_RETURN_FILTER_MH
+    @ForceInline
+    private static int returnFilter(MethodHandle errorHandle, int result) throws Throwable {
+        if (result >= 0) {
+            return result;
+        }
+        return -(int) errorHandle.invoke();
+    }
+
+    // Used reflectively via LONG_RETURN_FILTER_MH
+    @ForceInline
+    private static long returnFilter(MethodHandle errorHandle, long result) throws Throwable {
+        if (result >= 0) {
+            return result;
+        }
+        return -(int) errorHandle.invoke();
+    }
+
+    @SuppressWarnings("restricted")
+    private static MemorySegment malloc() {
+        final long address = UNSAFE.allocateMemory(SIZE);
+        return MemorySegment.ofAddress(address).reinterpret(SIZE);
+    }
+
+    private static void free(MemorySegment segment) {
+        UNSAFE.freeMemory(segment.address());
+    }
+
+    private static IllegalArgumentException illegalArgDoesNot(MethodHandle target, String info) {
+        return new IllegalArgumentException("The provided target " + target
+                + " does not " + info);
+    }
+
+    private static MethodHandle getAsIntHandle(StructLayout parent, MemoryLayout layout) {
+        final MethodHandle handle = MhUtil.findStatic(LOOKUP, "getStateAsInt", MethodType.methodType(int.class, VarHandle.class));
+        return handle.bindTo(parent.varHandle(MemoryLayout.PathElement.groupElement(layout.name().orElseThrow())));
+    }
+
+    // Used reflectively by `getAsIntHandle(MemoryLayout layout)`
+    private static int getStateAsInt(VarHandle handle) {
+        if (Thread.currentThread().isVirtual() && ContinuationSupport.isSupported()) {
```

Comment:
Shouldn't we pin for the entire duration of the native call + errno retrieval? Otherwise, can't we end up in a situation where we have a back to back native call where the second native call overwrites the errno state of the first?

Reply:
True. We need to rework the solution a bit as we do not want to pin virtual threads during system calls.

Reply:
A virtual thread can't unmount when there is a native frame on the continuation stack so it's pinned when in the syscall. For the errno capture then I think it will need to explicitly pin before the syscall, then unpin after the capture. The only access to the carrier local will be while pinned.